Testing Spatial/Temporal indexes

sherfert opened the pull request:
This PR adds unit tests and ITs around spatial and temporal indexes.

It also addresses some bugs that were uncovered by the new tests.

sherfert committed:
Add test for index populators and accessors (spatial/temporal)

This adds test for both unique and non-unique schema index
accessors and populators for all spatial and temporal types.

To fix tests it moves the sampling down from Spatial/Temporal
populators to the PartPopulators.

sherfert committed:
Removed IndexSamplerWrapper (which was just duplicated code)

sherfert committed:
Tests for IndexProviders

Adapt the lazy indexes to create at least one index on Populator.create
This ensures we can persist population error messages and have the
right initial status.

sherfert committed:
Use new types in index integration tests.

This includes random types for indexed node properties across
three ITs.

OliviaYtterbrink committed:
Handle range querys for line with area equal 0

Fixed SpatialSchemaIndexAccessorTest

sherfert committed:
Enforce a ratio of short-side to long-side search envelope max 1:100

To avoid bad performance when having very thin rectangles as spacial
index range queries, we ensure that the rectangle has at least a
ratio of 1:100 for the shorter and the longer sides of the rectangle/
cube. This will get more false positives that are handled during
post-filtering, but avoids having too many small search ranges for
the native index.

sherfert committed:
Merge branch '3.4' into 3.4-temporal-testing

sherfert committed:
Make LuceneIndexProvider use a subdirectory for lucene

If the LuceneIndexProvider provides access for lucene, temporal, and
spatial, it needs separate subfolders for each of them, otherwise
they will mess with each other's files.

sherfert committed:
Don't call IndexPopulator.close twice from test code

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DurationUniqueSchemaIndexAccessorTest.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    NativeSchemaIndexAccessor<DurationSchemaKey,NativeSchemaValue> makeAccessorWithSamplingConfig( IndexSamplingConfig samplingConfig ) throws IOException
+    {
+        TemporalIndexFiles.FileLayout<DurationSchemaKey> fileLayout = new TemporalIndexFiles.FileLayout<>( getIndexFile(), layout, ValueGroup.LOCAL_DATE_TIME );
```
Copy pasta: `ValueGroup.LOCAL_DATE_TIME`, should be `ValueGroup.DURATION`

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DurationLayoutTestUtil.java:
```diff
@@ -0,0 +1,106 @@
+{
+    private static final DurationValue[] ALL_EXTREME_VALUES = new DurationValue[]
+    {
+            DurationValue.duration( -999999999L * 12, 0, 0, 0),
```
If you want to test the extreme values for duration, you should multiply the values in these four lines with 2. This because people can write e.g. `duration.between(LocalDate.of( -999999999, 1, 1), LocalDate.of( 999999999, 12, 31)) ` .

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialLayoutTestUtil.java:
```diff
@@ -0,0 +1,118 @@
+            pointValue = Values.pointValue( crs, x, y );
+            compareValue = curve.derivedValueFor( pointValue.coordinate() );
+        }
+        while ( !uniqueCompareValues.add( compareValue ) );
```
I'm not entirely sure how the uniqueness constraint for Points was implemented, but does this mean that we will disregard points which are unique but happens to have the same 1D-value as an existing point? For me that sounds like an important case to test.

sherfert replied:
You're right. Unfortunately the test framework only exercises that part of the code, without the subsequent filtering that is needed for Spatial. That means if we allow Points that map to the same 1D-value here, the tests are failing because they do not expect to get more than one result for a query, but get more. The framework is not really designed for lossy indexes, but still provides _some_ nice coverage for Spatial.

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialUniqueSchemaIndexPopulatorTest.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    NativeSchemaIndexPopulator<SpatialSchemaKey,NativeSchemaValue> createPopulator( IndexSamplingConfig samplingConfig )
+    {
+        fileLayout = new SpatialIndexFiles.SpatialFileLayout( CoordinateReferenceSystem.WGS84, settings, super.getIndexFile() );
```
Maybe use your final variable `crs` here instead of `CoordinateReferenceSystem.WGS84`

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialNonUniqueSchemaIndexPopulatorTest.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    NativeSchemaIndexPopulator<SpatialSchemaKey,NativeSchemaValue> createPopulator( IndexSamplingConfig samplingConfig )
+    {
+        fileLayout = new SpatialIndexFiles.SpatialFileLayout( CoordinateReferenceSystem.WGS84, settings, super.getIndexFile() );
```
Maybe use your final variable `crs` here instead of `CoordinateReferenceSystem.WGS84`

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java:
```diff
@@ -697,7 +697,7 @@ public void shouldNotSeeFilteredEntries() throws Exception
         // when
         NodeValueIterator iter = new NodeValueIterator();
-        IndexQuery.ExactPredicate filter = IndexQuery.exact( 0, valueOf( updates[1] ) );
+        IndexQuery.ExactPredicate filter = IndexQuery.exact( 0, updates[1].values()[0].asObjectCopy() );
```
Why not using `valueOf( updates[1] ).asObjectCopy()` ?

sherfert replied:
Yeah, that's right.

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessorTest.java:
```diff
@@ -730,7 +730,7 @@ private Value generateUniqueValue( IndexEntryUpdate<SchemaIndexDescriptor>[] upd
     private Object valueOf( IndexEntryUpdate<SchemaIndexDescriptor> update )
     {
-        return update.values()[0].asObject();
+        return update.values()[0];
```
The method can now have the return type `Value`, which is more specific than `Object`

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulatorTest.java:
```diff
@@ -599,7 +598,8 @@ private void verifyUpdates( Iterator<IndexEntryUpdate<SchemaIndexDescriptor>> in
     {
         int size = 1000;
-        try ( StoreChannel storeChannel = fs.create( indexFile ) )
+        fs.mkdirs( getIndexFile().getParentFile() );
```
Out of curiosity, why was this addition necessary?

sherfert replied:
The structure of the indexes somehow got deeper (with a folder for the provider (e.g. lucene-1.0) and a subfolder for the sub-provider (e.g. temporal-1.0). Creating the file of these deeper nested directories requires creating the directories first. This was not true for Number/string/lucene, since they don't have sub-providers.

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/UniqueLayoutTestUtil.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override
+    IndexQuery rangeQuery( Object from, boolean fromInclusive, Object to, boolean toInclusive )
```
This method (+ its super method in `LayoutTestUtil` and its sub-methods) could change the signature to `IndexQuery rangeQuery( Value from, boolean fromInclusive, Value to, boolean toInclusive )`. See also comment about `valueOf`.

sherfert replied:
That's nice!

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/LayoutTestUtil.java:
```diff
@@ -54,7 +54,10 @@
-    protected abstract double fractionDuplicates();
+    protected double fractionDuplicates()
+    {
+        return 0.1;
```
`StringNonUniqueLayoutTestUtil` and `NumberNonUniqueLayoutTestUtil` override this method with exact the same implementation as here so those methods can be deleted.

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexProvider.java:
```diff
@@ -140,8 +141,15 @@ public InternalIndexState getInitialState( long indexId, SchemaIndexDescriptor d
+        {
+            monitor.failedToOpenIndex( indexId, descriptor, "Requesting re-population.",
+                    new NoSuchFileException( spatialIndexFiles.forCrs( CoordinateReferenceSystem.WGS84 ).indexFile.getAbsolutePath() ) );
+            return InternalIndexState.POPULATING;
```
Shouldn't we get `InternalIndexState.FAILED` after the exception on the line above? Same question also for  `TemporalIndexProvider`

sherfert replied:
Not according to this test:

```
@Test
    public void shouldReportInitialStateAsPopulatingIfIndexDoesntExist()
    {
        // given
        provider = newProvider();

        // when
        InternalIndexState state = provider.getInitialState( indexId, descriptor() );

        // then
        assertEquals( InternalIndexState.POPULATING, state );
        logging.assertContainsLogCallContaining( "Failed to open index" );
    }
```

It seems to be consistent with the other indexes, but I agree it's weird.

sherfert committed:
Moved rectangle padding to 1:100 to only maxDepth

instead of the actual search space.

Lojjs commented on community/spatial-index/src/test/java/org/neo4j/gis/spatial/index/EnvelopeTest.java:
```diff
@@ -106,6 +107,49 @@ public void shouldHandleIntersectionsIn2D()
+        // Expected to change
+        to = new double[] {100, 0.1, 12, 0.01};
+        Envelope envelope2 = new Envelope( from, to ).withSideRatioNotTooSmall();
+        double[] expectedTo2 = new double[] {100, 1, 12, 1};
```
In both this and the test above all changed values are expected to become 1, due to the highest diff being 100. Maybe we should test with some other value as well.

Lojjs reviewed (changes requested):
Good with more testing :) Some mixed comments though

sherfert committed:
Fix failing test by adapting to new behavior (and add some new).

Change max envelope side ratio to 1 000 000 000, which ends
up going to level 16 (instead of 4).

sherfert committed:
Fix subdirectories with Lucene and Spatial Temporal BW compatible

Lucene's folder cannot be moved for backwards compatibility reasons.
Thus, we keep it as /schema/index/lucene/1/... and keep temporal
and spatial not at the top level, but as a sibling like:
/schema/index/lucene-1.0/1/temporal-1.0/...
/schema/index/lucene-1.0/1/spatial-1.0/...

sherfert committed:
Fix all the review comments.

sherfert committed:
Make test wait for index to come online!

sherfert committed:
Changing value so that tests do not run out of heap space.

chrisvest commented on community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java:
```diff
@@ -174,6 +175,20 @@ private int countIndexedNodes( GraphDatabaseAPI db, Label label, String... keys
+                }
+                while ( state != InternalIndexState.ONLINE && runningFor < timeout );
+
+
```
There is also `db.schema().awaitIndexesOnline(…)`.

sherfert replied:
I think I don't have access to that from this IT, do I?

sherfert replied:
Yeah, let me use that one, I haven't seen that the GraphDatabaseAPI was in scope

sherfert replied:
I tried, but somehow that does not actually seem to wait for that index...
